tests/utils/managed_utils/mod.rs

```diff
@@ -0,0 +1,583 @@
+        let key_package = client.get_fresh_key_package(ciphersuite)?;
+        println!("Storing key package with hash: {:?}", key_package.hash());
+        self.waiting_for_welcome
+            .borrow_mut()
+            .insert(key_package.hash(), client.identity.clone());
+        Ok(key_package)
+    }
+
+    /// Deliver a Welcome message to group `group` to the intended recipients.
+    pub fn deliver_welcome(&self, welcome: Welcome, group: &Group) -> Result<(), SetupError> {
+        let clients = self.clients.borrow();
+        for egs in welcome.secrets() {
+            println!(
+                "Trying to get key package with hash: {:?}",
+                egs.key_package_hash
+            );
+            let client_id = match self
+                .waiting_for_welcome
+                .borrow_mut()
+                .remove(&egs.key_package_hash)
+            {
+                Some(id) => id,
+                None => return Err(SetupError::NoFreshKeyPackage),
+            };
+            let client = clients.get(&client_id).unwrap().borrow();
+            client.join_group(
+                group.group_config.clone(),
+                welcome.clone(),
+                Some(group.public_tree.clone()),
+            )?;
+        }
+        Ok(())
+    }
+
+    /// Distribute a set of messages sent by the sender with identity
+    /// `sender_id` to their intended recipients in group `Group`. This function
+    /// also verifies that all members of that group agree on the public tree.
+    pub fn distribute_to_members(
+        &self,
+        // We need the sender to know a group member that we know can not have
+        // been removed from the group.
+        sender_id: &Vec<u8>,
+        group: &mut Group,
+        messages: Vec<MLSMessage>,
+    ) -> Result<(), ClientError> {
+        let clients = self.clients.borrow();
+        println!("Distributing and processing messages...");
+        // Distribute message to all members.
+        for (index, member_id) in &group.members {
+            println!("Index: {:?}, Id: {:?}", index, member_id);
+            let member = clients.get(member_id).unwrap().borrow();
+            member.receive_messages_for_group(&group.group_id, messages.clone())?;
+        }
+        // Get the current tree and figure out who's still in the group.
+        let sender = clients.get(sender_id).unwrap().borrow();
+        let sender_groups = sender.groups.borrow();
+        let sender_group = sender_groups.get(&group.group_id).unwrap();
+        group.members = sender
+            .get_members_of_group(&group.group_id)?
+            .iter()
+            .map(|(index, cred)| (index.clone(), cred.identity().clone()))
+            .collect();
+        group.public_tree = sender_group.export_ratchet_tree();
+        drop(sender_group);
+        drop(sender_groups);
+        println!("Group members after distribution:");
+        // Check that the group states of all members match.
```

Comment:
Why? I can see that this is a nice thing to check in many cases. But there are at least as many interesting cases where this really shouldn't be the case.

Reply:
Fair point. Maybe it would be worth passing in a switch that enables/disables this behaviour or moving it to an external function of `Group`.

Reply:
Moved it to an external function that can be called independently.